### ISSUE TITLE
[BUG] Inconsistency with signing sub domains - vault not working

### ISSUE THREAD

**Issue by sseide** [rspamd/rspamd#4865]
### Prerequisites

* [x] Put an X between the brackets on this line if you have done all of the following:
    * Read about bug reporting in general: https://rspamd.com/doc/faq.html#how-to-report-bugs-found-in-rspamd
    * Enabled relevant debugging logs: https://rspamd.com/doc/faq.html#how-to-debug-some-module-in-rspamd 
    * Checked the FAQs about Core files in case of fatal crash: https://rspamd.com/doc/faq.html#how-to-figure-out-why-rspamd-process-crashed
    * Tried ASAN package and obtained the ASAN report (if possible): https://rspamd.com/doc/faq.html#asan-builds 
    * Checked that your issue isn't already filed: https://github.com/issues?utf8=%E2%9C%93&q=is%3Aissue+user%3Arspamd
    * Checked that there is not already an experimental package or master branch

**Describe the bug**

rspamd using DKIM signing with Vault and subdomains does not work. Subdomains are not signed.
And there is no fallback available to define this subdomains outside the fault.

But the vault being used exclusivly without the possibility to add additional domains with local keys is another issue i think.

I verified it with adding multiple additional debug logging inside the file `lualib/lua_dkim_tools.lua` at the functions `prepare_dkim_signing` and `sign_using_vault` to understand the data flow. I cannot debug the process, therefor more logging.

Excerpt from log with my addition log messages:
```
2024-03-08 16:16:51 #6(rspamd_proxy) <8f92a1>; lua; spf.lua:186: skip SPF checks for local networks and authorized users
2024-03-08 16:16:51 #6(rspamd_proxy) <8f92a1>; lua; dmarc.lua:349: skip DMARC checks as either SPF or DKIM were not checked
2024-03-08 16:16:51 #6(rspamd_proxy) <8f92a1>; dkim_signing; lua_dkim_tools.lua:166: mail is from local address
2024-03-08 16:16:51 #6(rspamd_proxy) <8f92a1>; dkim_signing; lua_dkim_tools.lua:382: use domain(header) for signature: sub.example.com
2024-03-08 16:16:51 #6(rspamd_proxy) <8f92a1>; dkim_signing; lua_dkim_tools.lua:402: final DKIM domain: sub.example.com
2024-03-08 16:16:51 #6(rspamd_proxy) <8f92a1>; dkim_signing; lua_dkim_tools.lua:403: start sanity checks for mail from: sub.example.com
2024-03-08 16:16:51 #6(rspamd_proxy) <8f92a1>; dkim_signing; lua_dkim_tools.lua:438: check DKIM sign by vault for domain: sub.example.com
2024-03-08 16:16:51 #6(rspamd_proxy) <8f92a1>; dkim_signing; lua_dkim_tools.lua:441: found vault DKIM domain: sub.example.com
2024-03-08 16:16:51 #6(rspamd_proxy) <8f92a1>; lua; once_received.lua:99: Skipping once_received for authenticated user or local network
2024-03-08 16:16:51 #6(rspamd_proxy) <8f92a1>; lua; greylist.lua:331: Score too low - skip greylisting
2024-03-08 16:16:51 #6(rspamd_proxy) <8f92a1>; proxy; rspamd_stat_classifier_is_skipped: learn condition for classifier bayes returned: already in class ham; probability 100.00%; skip classifier
```

Explaining the flow at the master branch code:
https://github.com/rspamd/rspamd/blob/master/lualib/lua_dkim_tools.lua

* `local function prepare_dkim_signing(N, task, settings)`
   it prepares everything and goes into the  `if settings.use_vault then` (line 461)
   Than it calls `if settings.vault_domains:get_key(dkim_domain) then` with `dkim_domain=sub.example.com`
   For some reason this method returns true even if this exact domain is not configured for using the vault - the vault is only configured for `example.com`
  Therefor this method returns 
  ```
  return true, {
          domain = 'sub.example.com',
          vault = true,
        }
   ```

* Later on the function `sign_using_vault` (line 618) is called
  It constructs the vault url with the subdomain - querying for a key of not the primary domain but the sub domain.
  line 622:
  `local full_url = string.format('%s/v1/%s/%s', settings.vault_url, settings.vault_path or 'dkim', selectors.domain)`
  => `https://..../v1/dkim/sub.example.com`

  The vault responses with an object containing no selector ( `{"data":{"selectors":[]}}` ). Now the following rspamd code iterates 'selectors' but does not find anything useful and silently fails not signing the mail leaving the admin with big question marks whats going on...
The filter selector function here does not handle the case if no valid selector is found.

### Steps to Reproduce

1. Use two domains: "example.com" and "sub.example.com"
2. Configure rspamd to use "example.com" from the fault
   ```
   dkim_signing {
     vault_domains [
        "example.com",
     ]
    use_vault = true;
    ...
  } 
3. Send mails through rspamd with a sender address of e.g. `user@example.com` - it gets signed correctly.
    Another mail with `user@sub.example.com` is not signed.

**Expected behavior**
Current behavoiur falsly tries to query a signing key from vault and therefor cannot sign the message. It does not check fallbacks or other configured signing tables. It does not print any error message or similar at all that it tries to sign but actually does not (as there are no keys returned). Therefor an admin does not have any clues WHY signing fails.

### Versions

Rspamd daemon version 3.2

Verified inside Github code of latest version too (master branch)

### Additional Information

I am not sure what the best behaviour is in this regards - not sign or fallback to primary domain and sign with this key.
But current implementation should be enhanced as it does not give any clue that/why siging fails.

**Comment by vstakhov** [comment]
> * For some reason this method returns true even if this exact domain is not configured for using the vault - the vault is only configured for `example.com`

Hum, this map is just a glob. So if there is `example.com` it should not match, but if there is `*example.com` it will match...

**Comment by vstakhov** [comment]
I have created a PR to address some of the issues in the ticket above. Thanks for the detailed description. I will appreciate if you can answer my question about `vault_domains` map as well.

**Comment by sseide** [comment]
Hi - thanks for the fast answer. The logging update will help definitely.

For the `vault_domains` map - i have no wildcard configured, only one TLD there. But i have tested it inside live system only with older version 3.2, not current. I do not known (and do not understand LUA enough) to check if there has things changed up to current version. Checked with master only common code function to see if important changes are inside the directly involve lua files regarding dkim key selection and signing.

From my point of view two different questions remain:

1.  why is the `settings.vault_domains:get_key(dkim_domain)` check true and the code enters the `if` condition?

2. There are multiple ways to configure DKIM singing domains and keys. Documentation(?) is a bit unclear about what ways can be used in parallel and what variants are exclusive.
At first i tried to add signing of the sub domain via `settings.domains` config key (full config below - for technical reasons not able to add key to fault) and was wondering why it is not used. Now looking into the code booth seems to be exclusive:
If `use_vault=true` check at line https://github.com/rspamd/rspamd/blob/eb67e5572bef87082c2cbabb6f033e39bfc04659/lualib/lua_dkim_tools.lua#L461 is true the function ends here no matter if acceptable key is found or not and code below like (e.g. `settings.domain`) is never evaluated. https://github.com/rspamd/rspamd/blob/eb67e5572bef87082c2cbabb6f033e39bfc04659/lualib/lua_dkim_tools.lua#L481
OTOH using dkim `settings.signing_table` can be used together with vault. https://github.com/rspamd/rspamd/blob/eb67e5572bef87082c2cbabb6f033e39bfc04659/lualib/lua_dkim_tools.lua#L257

So - either code should be modified to check all possible ways to configure DKIM signing or documentation should be updated to state what can be used in conjunction and what not (probably the easier task ;-) ) But this is up to your decision which way to go. Booth are fine from my point of view (at least as long as vault is not completely exclusive).

```
dkim_signing {
    allow_hdrfrom_mismatch = false;
    allow_username_mismatch = true;
    vault_domains [
        "example.com",
    ]
    symbol = "DKIM_SIGNED";
    sign_authenticated = true;
    selector = "dkim";
    allow_envfrom_empty = true;
    allow_hdrfrom_multiple = false;
    sign_networks [
        "127.1.2.3",
    ]
    sign_local = true;
    domain {
        sub.example.com {
            selector = "dkim";
            path = "/etc/rspamd/override.d/dkim/sub.example.com.dkim.key";
        }
    }
    vault_token = "my-secret-token";
    try_fallback = false;
    use_redis = false;
    key_prefix = "DKIM_KEYS";
    use_esld = false;
    use_vault = true;
    vault_url = "http://192.168.2.2/internal/rspamd/vault";
    use_domain = "header";
}
dmarc {}
dkim {
    time_jitter = 21600;
    trusted_only = false;
    skip_multi = false;
    dkim_cache_size = 2000;
    dkim_cache_expire = 86400;
}
